### DISCUSSION TITLE
Custom Animation

### DISCUSSION THREAD

**Discussion by JasonBSteele** [winleafs/Winleafs#198] [Show and tell]
Hi,

I have added the ability to create Custom Animation Effects. A video here demos this feature:
https://linksharing.samsungcloud.com/contents/view?contentsToken=1632160387343LSD7Bj9&currentIndex=1&linkUrlVersion=V1

Are you interested in a PR to merge this feature in?

**Issues I encountered**
I have used the name "CustomEffect" again because that is how the API refers to them but this term is already used to refer to "local" or "external" effects that get sent to device (single colour, screenmirror, etc.). I am happy to change the name to something else, or change the name already being used to something like "LocalEffect".

I had to add a package called RestClient.Serialisation.Newtonsoft because the default JSON serialisation used by RestClient ignores the [JsonProperty] attributes which specify the camelCase names for the properties. The Nanoleaf device was returning an error without this.

I haven't added the translations for the Netherlands yet, I can use Google Translate, or add them as English - let me know which you would prefer.

**More to consider**
I used the PercentageProfile window that was already in the solution as a basis for this. I have left the Window fixed at the same size, we may want to consider making it resizable, but then of course it feels less like a dialog.

The "Save Progress" button saves the current animation to the settings. You can only have one per device. We may want to consider supporting more than one by naming each with what the animation name will be, and have the user select the one they want to work on.

The animation only allows for one transition time to be specified between frames. The API allows the frames for a panel and the transition times of those frames to be set independently.

**Comment by bartdebever** [comment]
Hey there,

That's amazing to see! I've for sure talked about this idea before but never found the time to implement it. Looks really good!

We'd love to see a PR. No need to worry about the Dutch translations, I can do those.
RestClient.Serialisation.Newtonsoft is fine, I'm pretty sure I've used it in other projects as well after running into this project.
Great use of the PercentageProfile as well, was kind of designed for this purpose.

The review will go through the standard procedure with a code review and a functional/UI review. When you make the PR feel free to tag me as a reviewer and I'll communicate with @StijnOostdam where needed. If there are too personal things within the review or other things you don't feel comfortable doing, I'll be glad to talk over them or take them over.

Thank you for the great work and glad you see potential in the project!

**Comment by JasonBSteele** [comment]
Ok, that's great! I'll have a tidy up and sort that out over the next few days.
